odoo/fields.py

```diff
@@ -3047,6 +3058,518 @@ def _record_ids_per_res_model(self, records):
+        # convert_to_read is called, and the properties definition are fetched
+        # on the old parent, breaking the onchange
+        if not isinstance(value, list):
+            value = super().convert_to_onchange(value, record, names)
+
+        # add default values
+        for property_definition in value:
+            if property_definition.get('value') is None:
+                property_definition['value'] = property_definition.get('default')
+        return value
+
+    def _get_properties_definition(self, record):
+        parent = record[self.parent_record]
+        if not parent:
+            return
+
+        properties_definition = parent[self.parent_field]
+        return properties_definition
+
+    @classmethod
+    def _list_add_display_name(cls, values_list, env, value_key='value'):
+        """Add the name_get for each many2one / many2many properties.
+
+        :param values_list: List of properties definition and values
+        :param env: environment
+        """
+        for property_definition in values_list:
+            property_type = property_definition.get('type')
+            property_value = property_definition.get(value_key)
+            property_model = property_definition.get('model')
+            if not property_model:
+                continue
+
+            if property_type == 'many2one' and isinstance(property_value, int):
+                property_value = env[property_model].browse(property_value).name_get()
+                property_definition[value_key] = property_value[0] if property_value else False
+
+            elif property_type == 'many2many' and is_list_of(property_value, int):
+                property_value = env[property_model].browse(property_value).name_get()
+                property_definition[value_key] = property_value or False
+
+        return values_list
+
+    @classmethod
+    def _validate_values(cls, values_list, value_key='value'):
+        """Remove any incoherent properties values.
+
+        :param values_list: List of properties definition with value
+        :param value_key: The dict key which contains the values to validate
+        """
+        for values in values_list:
+            property_value = values.get(value_key)
+            if not property_value:
+                continue
+
+            property_type = values.get('type')
+            if property_type == 'many2one' and not isinstance(property_value, int):
+                values[value_key] = False
+
+            elif property_type == 'many2many':
+                if not is_list_of(property_value, int):
+                    values[value_key] = False
+                elif len(property_value) != len(set(property_value)):
+                    # remove duplicated value and preserve order
+                    values[value_key] = list(dict.fromkeys(property_value))
+
+        return values_list
+
+    @classmethod
+    def _parse_json_types(cls, values_list, env):
```

Comment:
"_parse_json_types" looks like a more elaborate "_validate_values", can't we merge both?